# 1. Go 语言错误处理语句

## 1.1 err 到底是什么？

error 内置接口类型是表示错误条件的常规接口，nil值表示无错误。

```go
// The error built-in interface type is the conventional interface for
// representing an error condition, with the nil value representing no error.
type error interface {
	Error() string
}
```

下面是一个会产生错误的代码示例：

```go
package main

import (
   "bufio"
   "fmt"
   "os"
)

type IntGen func() int

func fibonacci() IntGen {
   // 斐波那契数列，返回一个intGen类型
   a, b := 0, 1
   return func() int {
      a, b = b, a+b
      return a
   }
}

func writeFile(filename string) {
   // 我们在这里来看看err到底是什么作用?
   file, err := os.OpenFile(filename, os.O_EXCL | os.O_CREATE, 0666)
   if err != nil 
  		// 如果不加Error() 方法的调用也是可以的，Println会自动拿到其中的字符串
      fmt.Println(err.Error()) // open fib.txt: file exists
      return
   }
   defer file.Close()
   writer := bufio.NewWriter(file)
   defer writer.Flush()
   f := fibonacci()
   for i := 0; i < 20; i++ {
      fmt.Fprintln(writer, f())
   }

}

func main() {
   writeFile("fib.txt")
}
```

## 1.2 查看实际的错误类型 [PathError为例]

接着我们看openFile的源代码，从而从注释上可以知道实际返回的错误类型是哪一个类型：下面有这样一行注释：If there is an error, it will be of type *PathError，所以我们就能够知道实际返回的类型就是PathError类型。

```go
// OpenFile is the generalized open call; most users will use Open
// or Create instead. It opens the named file with specified flag
// (O_RDONLY etc.). If the file does not exist, and the O_CREATE flag
// is passed, it is created with mode perm (before umask). If successful,
// methods on the returned File can be used for I/O.
// If there is an error, it will be of type *PathError.
func OpenFile(name string, flag int, perm FileMode) (*File, error) {
   testlog.Open(name)
   f, err := openFileNolog(name, flag, perm)
   if err != nil {
      return nil, err
   }
   f.appendMode = flag&O_APPEND != 0

   return f, nil
}
```

我们将类型转换为PathError类型：从而可以运用其中的属性，如pathError.Op, pathError.Path pathError.Err

```go
package main

import (
   "bufio"
   "fmt"
   "os"
)

type IntGen func() int

func fibonacci() IntGen {
   // 斐波那契数列，返回一个intGen类型
   a, b := 0, 1
   return func() int {
      a, b = b, a+b
      return a
   }
}

func writeFile(filename string) {
   // 我们在这里来看看err到底是什么作用?
   file, err := os.OpenFile(filename, os.O_EXCL|os.O_CREATE, 0666)
   if err != nil {
      // 注意由于实现的方法是指针接收者，所以这里强制类型转换为*os.PathError
      if pathError, ok := err.(*os.PathError); !ok {
         panic(err)
      } else {
         fmt.Println("操作: ", pathError.Op, " 路径: ", pathError.Path, " 错误：", pathError.Err)
         // 操作:  open  路径:  fib.txt  错误： file exists
      }
      return
   }
   defer file.Close()
   writer := bufio.NewWriter(file)
   defer writer.Flush()
   f := fibonacci()
   for i := 0; i < 20; i++ {
      fmt.Fprintln(writer, f())
   }
}

func main() {
   writeFile("fib.txt")
}
```